File: app/src/main/java/de/haukesomm/sokoban/game/moving/validation/OutOfBoundsPreventingMoveValidator.java
package de.haukesomm.sokoban.game.moving.validation;

import de.haukesomm.sokoban.game.Direction;
import de.haukesomm.sokoban.game.Entity;
import de.haukesomm.sokoban.game.GameState;
import de.haukesomm.sokoban.game.moving.MoveValidatorStatus;

import java.util.Collection;

public class OutOfBoundsPreventingMoveValidator extends AbstractMoveValidator {

    @Override
    public Collection<MoveValidatorStatus> check(GameState state, Entity entity, Direction direction) {
        var nextPosition = entity.position().nextInDirection(direction);
        if (nextPosition.x() < 0 ||
                nextPosition.y() < 0 ||
                nextPosition.x() >= state.getMapWidth() ||
                nextPosition.y() >= state.getMapHeight()) {
            // TODO: Use logger with warning level
            System.err.println("Warning: Level is not fully surrounded by walls!");
            return singleResult(MoveValidatorStatus.IMPOSSIBLE);
        }

        return singleResult(MoveValidatorStatus.POSSIBLE);
    }
}

File: core/src/commonMain/kotlin/de/haukesomm/sokoban/core/moving/validation/OutOfBoundsPreventingMoveValidator.kt
package de.haukesomm.sokoban.core.moving.validation

import de.haukesomm.sokoban.core.Direction
import de.haukesomm.sokoban.core.Entity
import de.haukesomm.sokoban.core.GameState

class OutOfBoundsPreventingMoveValidator : AbstractMoveValidator() {

    override fun check(state: GameState, entity: Entity, direction: Direction): Collection<MoveValidatorStatus> {
        val (x, y) = entity.position.nextInDirection(direction)
        if (x < 0 || y < 0 || x >= state.getMapWidth() || y >= state.getMapHeight()) {
            // TODO: Use logger with warning level
            println("Warning: Level is not fully surrounded by walls!")
            return singleResult(MoveValidatorStatus.IMPOSSIBLE)
        }
        return singleResult(MoveValidatorStatus.POSSIBLE)
    }
}
